File: BTFViewer/web/src/utils/findingsTriage.js
/**
 * Findings triage queue helpers.
 * Keep in sync with btf_viewer_pkg/findings_triage.py.
 */

import { evidenceStrengthBadge, normalizeEvidenceStrength } from './evidenceStrength.js'

const SEVERITY_ORDER = { error: 0, warning: 1, info: 2, ask: 3 }

export const FINDING_CATEGORIES = [
  'migration', 'blocking', 'deadline', 'load', 'jitter',
  'execution', 'dispatch', 'general',
]

export const SORT_SEVERITY = 'severity'
export const SORT_EVIDENCE = 'evidence'
export const SORT_TITLE = 'title'
export const SORT_CATEGORY = 'category'
export const SORT_KEYS = [SORT_SEVERITY, SORT_EVIDENCE, SORT_TITLE, SORT_CATEGORY]
export const SORT_LABELS = {
  [SORT_SEVERITY]: 'Severity',
  [SORT_EVIDENCE]: 'Evidence strength',
  [SORT_TITLE]: 'Title',
  [SORT_CATEGORY]: 'Category',
}

export function findingCategory(finding) {
  const blob = `${finding.title || ''} ${finding.text || ''}`.toLowerCase()
  if (/migrat|thrash|bounce/.test(blob)) return 'migration'
  if (/block|mutex|wait/.test(blob)) return 'blocking'
  if (/deadline|budget/.test(blob)) return 'deadline'
  if (/load|balance|gini/.test(blob)) return 'load'
  if (/jitter|period/.test(blob)) return 'jitter'
  if (/wcet|execution|cpu/.test(blob)) return 'execution'
  if (/dispatch|latency/.test(blob)) return 'dispatch'
  return 'general'
}

export function findingEvidenceStrength(finding) {
  const ev = finding.evidence || []
  if (ev.some(e => e && typeof e === 'object' && e.time != null)) return 'direct'
  if (ev.length) return 'derived'
  return 'estimated'
}

function defaultCheckNext(finding) {
  const cat = findingCategory(finding)
  const mapping = {
    migration: 'Open Core Migrations and check load balance first.',
    blocking: 'Inspect Blocking Time and Mutex Blocking around the cited time.',
    deadline: 'Open Deadline / CPU Budget and Response Time.',
    load: 'Open Load Balance and Task × Core.',
    jitter: 'Open Period / Jitter and Recurring Patterns.',
    execution: 'Open Execution and Worst Events for Max/outliers.',
    dispatch: 'Open Dispatch latency and Execution.',
  }
  return mapping[cat] || 'Open Statistics for the related metric and verify on the timeline.'
}

export function enrichFindingCard(finding) {
  const title = String(finding.title || 'Finding')
  const text = String(finding.text || '')
  const strength = findingEvidenceStrength(finding)
  const badge = evidenceStrengthBadge(strength)
  const evLines = []
  for (const ev of finding.evidence || []) {
    if (!ev || typeof ev !== 'object') continue
    const label = String(ev.label || ev.text || 'evidence')
    evLines.push(ev.time != null ? `${label}: jump:${ev.time}` : label)
  }
  return {
    ...finding,
    category: findingCategory(finding),
    evidence_strength: strength,
    evidence_strength_label: badge.label,
    observation: title,
    evidence_text: evLines.length ? evLines.join('; ') : 'No timed evidence yet.',
    why_it_matters: text || 'May indicate a timing or scheduling problem in the current Scope.',
    check_next: finding.check_next || defaultCheckNext(finding),
  }
}

export function sortFindingsTriage(items, { sortBy = SORT_SEVERITY } = {}) {
  let mode = String(sortBy || SORT_SEVERITY).trim().toLowerCase()
  if (!SORT_KEYS.includes(mode)) mode = SORT_SEVERITY
  const strengthRank = { direct: 0, derived: 1, estimated: 2, configured: 1 }
  const enriched = (items || []).filter(f => f && typeof f === 'object').map(f => enrichFindingCard({ ...f }))
  enriched.sort((a, b) => {
    const sa = SEVERITY_ORDER[String(a.severity || 'info').toLowerCase()] ?? 9
    const sb = SEVERITY_ORDER[String(b.severity || 'info').toLowerCase()] ?? 9
    const ea = strengthRank[normalizeEvidenceStrength(a.evidence_strength)] ?? 3
    const eb = strengthRank[normalizeEvidenceStrength(b.evidence_strength)] ?? 3
    const ta = String(a.title || '')
    const tb = String(b.title || '')
    const ca = String(a.category || '')
    const cb = String(b.category || '')
    if (mode === SORT_EVIDENCE) {
      if (ea !== eb) return ea - eb
      if (sa !== sb) return sa - sb
      return ta.localeCompare(tb)
    }
    if (mode === SORT_TITLE) {
      const cmp = ta.localeCompare(tb)
      if (cmp) return cmp
      if (sa !== sb) return sa - sb
      return ea - eb
    }
    if (mode === SORT_CATEGORY) {
      const cmp = ca.localeCompare(cb)
      if (cmp) return cmp
      if (sa !== sb) return sa - sb
      if (ea !== eb) return ea - eb
      return ta.localeCompare(tb)
    }
    if (sa !== sb) return sa - sb
    if (ea !== eb) return ea - eb
    return ta.localeCompare(tb)
  })
  return enriched
}

export function filterFindingsTriage(items, {
  severity = '', category = '', task = '', core = '',
  evidenceStrength = '', sortBy = SORT_SEVERITY,
} = {}) {
  const sevW = String(severity || '').trim().toLowerCase()
  const catW = String(category || '').trim().toLowerCase()
  const taskW = String(task || '').trim().toLowerCase()
  const coreW = String(core || '').trim().toLowerCase()
  const evW = String(evidenceStrength || '').trim().toLowerCase()
  return sortFindingsTriage(items, { sortBy }).filter(f => {
    if (sevW && String(f.severity || '').toLowerCase() !== sevW) return false
    if (catW && String(f.category || '').toLowerCase() !== catW) return false
    const blob = `${f.title || ''} ${f.text || ''} ${f.task || ''} ${f.core || ''}`.toLowerCase()
    if (taskW && !blob.includes(taskW)) return false
    if (coreW && !blob.includes(coreW)) return false
    if (evW && normalizeEvidenceStrength(f.evidence_strength) !== evW) return false
    return true
  })
}

export function findingFilterFacets(items) {
  const cats = new Set()
  const tasks = new Set()
  const cores = new Set()
  for (const f of sortFindingsTriage(items || [])) {
    cats.add(String(f.category || 'general'))
    const task = String(f.task || '').trim()
    if (task) tasks.add(task)
    const core = String(f.core || '').trim()
    if (core) cores.add(core)
    const blob = `${f.title || ''} ${f.text || ''}`
    const re = /\bCore[_ ]?\d+\b/gi
    let m
    while ((m = re.exec(blob))) cores.add(m[0].replace(/ /g, '_'))
  }
  return {
    categories: [...cats].sort(),
    tasks: [...tasks].sort((a, b) => a.localeCompare(b)),
    cores: [...cores].sort((a, b) => a.localeCompare(b)),
  }
}

export function groupFindingsByIncident(items, clusters = [], { group = true } = {}) {
  const findings = (items || []).filter(f => f && typeof f === 'object')
  if (!group || !findings.length) {
    return findings.map(f => ({ kind: 'finding', ...f }))
  }
  const incs = (clusters || []).filter(c => c && typeof c === 'object')
  const idToInc = new Map()
  const titleToInc = new Map()
  for (const inc of incs) {
    for (const fid of inc.finding_ids || []) {
      if (fid) idToInc.set(String(fid), inc)
    }
    for (const title of inc.findings || []) {
      if (title) titleToInc.set(String(title), inc)
    }
  }
  const buckets = new Map()
  const order = []
  const ungrouped = []
  for (const f of findings) {
    const fid = String(f.id || '')
    const title = String(f.title || f.observation || '')
    const inc = idToInc.get(fid) || titleToInc.get(title)
    if (!inc || Number(inc.count || 0) < 2) {
      ungrouped.push(f)
      continue
    }
    const cid = String(inc.id || '')
    if (!buckets.has(cid)) {
      buckets.set(cid, [])
      order.push(cid)
    }
    buckets.get(cid).push(f)
  }
  const rows = []
  for (const cid of order) {
    const members = buckets.get(cid) || []
    if (members.length < 2) {
      for (const m of members) rows.push({ kind: 'finding', ...m })
      continue
    }
    const inc = idToInc.get(String(members[0].id || '')) || {}
    const root = String(inc.root_suspect || 'mixed')
    rows.push({
      kind: 'header',
      incident_id: cid,
      label: `${cid} · ${root}`,
      count: members.length,
      finding_ids: members.map(m => String(m.id || '')),
    })
    for (const m of members) rows.push({ kind: 'finding', incident_id: cid, ...m })
  }
  for (const f of ungrouped) rows.push({ kind: 'finding', ...f })
  return rows
}

export function formatInvestigatePreview(finding, {
  scope = null, sectionId = '', sectionLabel = '',
  currentLimit = false, currentLo = null, currentHi = null,
} = {}) {
  if (!finding || typeof finding !== 'object') {
    return 'Select a finding to preview Investigate.'
  }
  const lines = ['Investigate will:']
  const sid = String(sectionId || '').trim()
  const slabel = String(sectionLabel || sid || 'related Statistics').trim()
  lines.push(`  • Open Statistics → ${slabel}`)
  if (scope && scope.lo != null && scope.hi != null) {
    const lo = Math.trunc(Number(scope.lo))
    const hi = Math.trunc(Number(scope.hi))
    const reason = String(scope.reason || 'recommended evidence window').trim()
    lines.push(`  • Place C1–C2 at ${lo}–${hi} (${reason})`)
    lines.push('  • Enable Limit to C1–Cn for Statistics / Findings')
    if (currentLimit && currentLo != null && currentHi != null) {
      lines.push(`  • Replaces current Scope ${Math.trunc(Number(currentLo))}–${Math.trunc(Number(currentHi))}`)
    } else if (currentLimit) {
      lines.push('  • Replaces the current cursor Scope')
    } else {
      lines.push('  • Scope is currently Full Trace (Limit off)')
    }
  } else {
    lines.push('  • Keep the current cursor Scope (no recommended window)')
  }
  lines.push('You can Undo the Scope change after Confirm.')
  return lines.join('\n')
}

/** Queue labels shown in the Analysis Findings strip (Done = reviewed). */
export const QUEUE_OPEN = 'open'
export const QUEUE_DONE = 'done'
export const QUEUE_CASE = 'case'
export const QUEUE_DISMISSED = 'dismissed'
export const QUEUE_IDS = [QUEUE_OPEN, QUEUE_DONE, QUEUE_CASE, QUEUE_DISMISSED]

export function defaultTriageState() {
  return { reviewed: [], dismissed: {}, case: [] }
}

export function normalizeTriageState(state) {
  const base = defaultTriageState()
  if (!state || typeof state !== 'object') return base
  const reviewed = (state.reviewed || []).map(x => String(x).trim()).filter(Boolean)
  const caseIds = (state.case || []).map(x => String(x).trim()).filter(Boolean)
  const dismissed = {}
  const raw = state.dismissed || {}
  if (raw && typeof raw === 'object') {
    for (const [k, v] of Object.entries(raw)) {
      const kid = String(k).trim()
      if (kid) dismissed[kid] = String(v || 'Dismissed')
    }
  }
  return { reviewed, dismissed, case: caseIds }
}

/** Primary queue for a finding: dismissed > case > done > open. */
export function findingQueueStatus(findingId, state = null) {
  const st = normalizeTriageState(state)
  const fid = String(findingId || '').trim()
  if (!fid) return QUEUE_OPEN
  if (st.dismissed?.[fid]) return QUEUE_DISMISSED
  if ((st.case || []).includes(fid)) return QUEUE_CASE
  if ((st.reviewed || []).includes(fid)) return QUEUE_DONE
  return QUEUE_OPEN
}

export function filterByQueue(items, state = null, { queue = QUEUE_OPEN } = {}) {
  let q = String(queue || QUEUE_OPEN).trim().toLowerCase() || QUEUE_OPEN
  if (!QUEUE_IDS.includes(q)) q = QUEUE_OPEN
  const st = normalizeTriageState(state)
  return (items || []).filter(f => {
    if (!f || typeof f !== 'object') return false
    return findingQueueStatus(f.id || '', st) === q
  })
}

export function queueCounts(items, state = null) {
  const st = normalizeTriageState(state)
  const counts = Object.fromEntries(QUEUE_IDS.map(id => [id, 0]))
  for (const f of items || []) {
    if (!f || typeof f !== 'object') continue
    const fid = String(f.id || '').trim()
    if (!fid) {
      counts[QUEUE_OPEN] += 1
      continue
    }
    counts[findingQueueStatus(fid, st)] += 1
  }
  return counts
}

export function applyTriageAction(state, findingId, action, { reason = '' } = {}) {
  const out = normalizeTriageState(state)
  const fid = String(findingId || '').trim()
  const act = String(action || '').trim().toLowerCase()
  if (!fid) return out
  if (act === 'reviewed' || act === 'done') {
    const reviewed = [...(out.reviewed || [])]
    if (!reviewed.includes(fid)) reviewed.push(fid)
    out.reviewed = reviewed
    const dismissed = { ...(out.dismissed || {}) }
    delete dismissed[fid]
    out.dismissed = dismissed
  } else if (act === 'unreviewed' || act === 'undo_done') {
    out.reviewed = (out.reviewed || []).filter(x => x !== fid)
  } else if (act === 'dismiss') {
    out.dismissed = { ...(out.dismissed || {}), [fid]: String(reason || 'Dismissed') }
    out.reviewed = (out.reviewed || []).filter(x => x !== fid)
    out.case = (out.case || []).filter(x => x !== fid)
  } else if (act === 'undismiss' || act === 'restore') {
    const dismissed = { ...(out.dismissed || {}) }
    delete dismissed[fid]
    out.dismissed = dismissed
  } else if (act === 'case') {
    const caseIds = [...(out.case || [])]
    if (!caseIds.includes(fid)) caseIds.push(fid)
    out.case = caseIds
    const dismissed = { ...(out.dismissed || {}) }
    delete dismissed[fid]
    out.dismissed = dismissed
  } else if (act === 'uncase') {
    out.case = (out.case || []).filter(x => x !== fid)
  }
  return out
}

/** Append Done / Case / Dismissed sections for Save as text. */
export function formatTriageAuditText(findings, state = null) {
  const st = normalizeTriageState(state)
  const byId = {}
  for (const f of findings || []) {
    if (!f || typeof f !== 'object') continue
    const fid = String(f.id || '').trim()
    if (fid) byId[fid] = f
  }
  const titleOf = (fid) => String((byId[fid] || {}).title || fid)
  const lines = []
  const done = st.reviewed || []
  const caseIds = st.case || []
  const dismissed = st.dismissed || {}
  if (done.length) {
    lines.push('Done:')
    for (const fid of done) lines.push(`  - ${titleOf(fid)} (id=${fid})`)
    lines.push('')
  }
  if (caseIds.length) {
    lines.push('In case:')
    for (const fid of caseIds) lines.push(`  - ${titleOf(fid)} (id=${fid})`)
    lines.push('')
  }
  if (Object.keys(dismissed).length) {
    lines.push('Dismissed:')
    for (const [fid, reason] of Object.entries(dismissed)) {
      lines.push(`  - ${titleOf(fid)} (id=${fid}): ${reason}`)
    }
    lines.push('')
  }
  return lines.join('\n').replace(/\n+$/, '')
}

File: BTFViewer/btf_viewer_pkg/findings_triage.py
"""Findings triage queue helpers.

Lockstep with ``web/src/utils/findingsTriage.js``.
"""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Sequence

from .evidence_strength import evidence_strength_badge, normalize_evidence_strength


SEVERITY_ORDER = {"error": 0, "warning": 1, "info": 2, "ask": 3}

FINDING_CATEGORIES = (
    "migration", "blocking", "deadline", "load", "jitter",
    "execution", "dispatch", "general",
)

# Sort keys for the Analysis Findings Sort control (default = severity).
SORT_SEVERITY = "severity"
SORT_EVIDENCE = "evidence"
SORT_TITLE = "title"
SORT_CATEGORY = "category"
SORT_KEYS = (SORT_SEVERITY, SORT_EVIDENCE, SORT_TITLE, SORT_CATEGORY)
SORT_LABELS = {
    SORT_SEVERITY: "Severity",
    SORT_EVIDENCE: "Evidence strength",
    SORT_TITLE: "Title",
    SORT_CATEGORY: "Category",
}


def finding_category(finding: Dict[str, Any]) -> str:
    blob = f"{finding.get('title') or ''} {finding.get('text') or ''}".lower()
    if "migrat" in blob or "thrash" in blob or "bounce" in blob:
        return "migration"
    if "block" in blob or "mutex" in blob or "wait" in blob:
        return "blocking"
    if "deadline" in blob or "budget" in blob:
        return "deadline"
    if "load" in blob or "balance" in blob or "gini" in blob:
        return "load"
    if "jitter" in blob or "period" in blob:
        return "jitter"
    if "wcet" in blob or "execution" in blob or "cpu" in blob:
        return "execution"
    if "dispatch" in blob or "latency" in blob:
        return "dispatch"
    return "general"


def finding_evidence_strength(finding: Dict[str, Any]) -> str:
    ev = finding.get("evidence") or []
    if any(isinstance(e, dict) and e.get("time") is not None for e in ev):
        return "direct"
    if ev:
        return "derived"
    return "estimated"


def enrich_finding_card(finding: Dict[str, Any]) -> Dict[str, Any]:
    """Observation / Evidence / Why / Check next card structure."""
    title = str(finding.get("title") or "Finding")
    text = str(finding.get("text") or "")
    strength = finding_evidence_strength(finding)
    badge = evidence_strength_badge(strength)
    ev_lines: List[str] = []
    for ev in finding.get("evidence") or []:
        if not isinstance(ev, dict):
            continue
        label = str(ev.get("label") or ev.get("text") or "evidence")
        t = ev.get("time")
        ev_lines.append(f"{label}: jump:{t}" if t is not None else label)
    return {
        **finding,
        "category": finding_category(finding),
        "evidence_strength": strength,
        "evidence_strength_label": badge["label"],
        "observation": title,
        "evidence_text": "; ".join(ev_lines) if ev_lines else "No timed evidence yet.",
        "why_it_matters": text or "May indicate a timing or scheduling problem in the current Scope.",
        "check_next": finding.get("check_next") or _default_check_next(finding),
    }


def _default_check_next(finding: Dict[str, Any]) -> str:
    cat = finding_category(finding)
    mapping = {
        "migration": "Open Core Migrations and check load balance first.",
        "blocking": "Inspect Blocking Time and Mutex Blocking around the cited time.",
        "deadline": "Open Deadline / CPU Budget and Response Time.",
        "load": "Open Load Balance and Task × Core.",
        "jitter": "Open Period / Jitter and Recurring Patterns.",
        "execution": "Open Execution and Worst Events for Max/outliers.",
        "dispatch": "Open Dispatch latency and Execution.",
    }
    return mapping.get(cat, "Open Statistics for the related metric and verify on the timeline.")


def sort_findings_triage(
    items: Sequence[Dict[str, Any]],
    *,
    sort_by: str = SORT_SEVERITY,
) -> List[Dict[str, Any]]:
    mode = str(sort_by or SORT_SEVERITY).strip().lower()
    if mode not in SORT_KEYS:
        mode = SORT_SEVERITY
    strength_rank = {"direct": 0, "derived": 1, "estimated": 2, "configured": 1}

    def key(f: Dict[str, Any]) -> tuple:
        sev = SEVERITY_ORDER.get(str(f.get("severity") or "info").lower(), 9)
        strength = normalize_evidence_strength(
            f.get("evidence_strength") or finding_evidence_strength(f))
        erank = strength_rank.get(strength, 3)
        title = str(f.get("title") or "")
        cat = str(f.get("category") or finding_category(f))
        if mode == SORT_EVIDENCE:
            return (erank, sev, title)
        if mode == SORT_TITLE:
            return (title.lower(), sev, erank)
        if mode == SORT_CATEGORY:
            return (cat.lower(), sev, erank, title)
        return (sev, erank, title)

    return sorted(
        [enrich_finding_card(dict(f)) for f in items if isinstance(f, dict)],
        key=key,
    )


def filter_findings_triage(
    items: Sequence[Dict[str, Any]],
    *,
    severity: str = "",
    category: str = "",
    task: str = "",
    core: str = "",
    evidence_strength: str = "",
    sort_by: str = SORT_SEVERITY,
) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    sev_w = str(severity or "").strip().lower()
    cat_w = str(category or "").strip().lower()
    task_w = str(task or "").strip().lower()
    core_w = str(core or "").strip().lower()
    ev_w = str(evidence_strength or "").strip().lower()
    for f in sort_findings_triage(items, sort_by=sort_by):
        if sev_w and str(f.get("severity") or "").lower() != sev_w:
            continue
        if cat_w and str(f.get("category") or "").lower() != cat_w:
            continue
        blob = (
            f"{f.get('title') or ''} {f.get('text') or ''} "
            f"{f.get('task') or ''} {f.get('core') or ''}"
        ).lower()
        if task_w and task_w not in blob:
            continue
        if core_w and core_w not in blob:
            continue
        if ev_w and normalize_evidence_strength(f.get("evidence_strength")) != ev_w:
            continue
        out.append(f)
    return out


def finding_filter_facets(items: Sequence[Dict[str, Any]]) -> Dict[str, List[str]]:
    """Distinct category / task / core values for filter controls."""
    cats: set = set()
    tasks: set = set()
    cores: set = set()
    for f in sort_findings_triage(items or []):
        cats.add(str(f.get("category") or "general"))
        task = str(f.get("task") or "").strip()
        if task:
            tasks.add(task)
        core = str(f.get("core") or "").strip()
        if core:
            cores.add(core)
        # Also scrape Core_N from text when no explicit field.
        blob = f"{f.get('title') or ''} {f.get('text') or ''}"
        for m in re.finditer(r"\bCore[_ ]?\d+\b", blob, re.I):
            cores.add(m.group(0).replace(" ", "_"))
    return {
        "categories": sorted(cats),
        "tasks": sorted(tasks, key=str.lower),
        "cores": sorted(cores, key=str.lower),
    }


def group_findings_by_incident(
    items: Sequence[Dict[str, Any]],
    clusters: Optional[Sequence[Dict[str, Any]]] = None,
    *,
    group: bool = True,
) -> List[Dict[str, Any]]:
    """Flat list of display rows: header rows + finding rows.

    Each row is ``{"kind": "header"|"finding", ...}``. Findings keep their
    fields; headers carry ``incident_id``, ``label``, ``count``.
    """
    findings = [f for f in (items or []) if isinstance(f, dict)]
    if not group or not findings:
        return [{"kind": "finding", **f} for f in findings]
    incs = [c for c in (clusters or []) if isinstance(c, dict)]
    id_to_inc: Dict[str, Dict[str, Any]] = {}
    title_to_inc: Dict[str, Dict[str, Any]] = {}
    for inc in incs:
        for fid in inc.get("finding_ids") or []:
            if fid:
                id_to_inc[str(fid)] = inc
        for title in inc.get("findings") or []:
            if title:
                title_to_inc[str(title)] = inc
    buckets: Dict[str, List[Dict[str, Any]]] = {}
    order: List[str] = []
    ungrouped: List[Dict[str, Any]] = []
    for f in findings:
        fid = str(f.get("id") or "")
        title = str(f.get("title") or f.get("observation") or "")
        inc = id_to_inc.get(fid) or title_to_inc.get(title)
        if not inc or int(inc.get("count") or 0) < 2:
            ungrouped.append(f)
            continue
        cid = str(inc.get("id") or "")
        if cid not in buckets:
            buckets[cid] = []
            order.append(cid)
        buckets[cid].append(f)
    rows: List[Dict[str, Any]] = []
    for cid in order:
        members = buckets.get(cid) or []
        if len(members) < 2:
            rows.extend({"kind": "finding", **m} for m in members)
            continue
        inc = id_to_inc.get(str(members[0].get("id") or "")) or {}
        root = str(inc.get("root_suspect") or "mixed")
        rows.append({
            "kind": "header",
            "incident_id": cid,
            "label": f"{cid} · {root}",
            "count": len(members),
            "finding_ids": [str(m.get("id") or "") for m in members],
        })
        for m in members:
            rows.append({"kind": "finding", "incident_id": cid, **m})
    rows.extend({"kind": "finding", **f} for f in ungrouped)
    return rows


def format_investigate_preview(
    finding: Optional[Dict[str, Any]] = None,
    *,
    scope: Optional[Dict[str, Any]] = None,
    section_id: str = "",
    section_label: str = "",
    current_limit: bool = False,
    current_lo: Optional[float] = None,
    current_hi: Optional[float] = None,
) -> str:
    """Human-readable preview of what Investigate will change."""
    if not isinstance(finding, dict):
        return "Select a finding to preview Investigate."
    lines = ["Investigate will:"]
    sid = str(section_id or "").strip()
    slabel = str(section_label or sid or "related Statistics").strip()
    lines.append(f"  • Open Statistics → {slabel}")
    if isinstance(scope, dict) and scope.get("lo") is not None and scope.get("hi") is not None:
        lo = int(scope["lo"])
        hi = int(scope["hi"])
        reason = str(scope.get("reason") or "recommended evidence window").strip()
        lines.append(f"  • Place C1–C2 at {lo}–{hi} ({reason})")
        lines.append("  • Enable Limit to C1–Cn for Statistics / Findings")
        if current_limit and current_lo is not None and current_hi is not None:
            lines.append(
                f"  • Replaces current Scope {int(current_lo)}–{int(current_hi)}"
            )
        elif current_limit:
            lines.append("  • Replaces the current cursor Scope")
        else:
            lines.append("  • Scope is currently Full Trace (Limit off)")
    else:
        lines.append("  • Keep the current cursor Scope (no recommended window)")
    lines.append("You can Undo the Scope change after Confirm.")
    return "\n".join(lines)


# Queue labels shown in the Analysis Findings strip (Done = reviewed).
QUEUE_OPEN = "open"
QUEUE_DONE = "done"
QUEUE_CASE = "case"
QUEUE_DISMISSED = "dismissed"
QUEUE_IDS = (QUEUE_OPEN, QUEUE_DONE, QUEUE_CASE, QUEUE_DISMISSED)


def default_triage_state() -> Dict[str, Any]:
    return {"reviewed": [], "dismissed": {}, "case": []}


def normalize_triage_state(state: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    base = default_triage_state()
    if not isinstance(state, dict):
        return base
    reviewed = [str(x).strip() for x in (state.get("reviewed") or []) if str(x).strip()]
    case = [str(x).strip() for x in (state.get("case") or []) if str(x).strip()]
    dismissed_raw = state.get("dismissed") or {}
    dismissed: Dict[str, str] = {}
    if isinstance(dismissed_raw, dict):
        for k, v in dismissed_raw.items():
            kid = str(k).strip()
            if kid:
                dismissed[kid] = str(v or "Dismissed")
    return {"reviewed": reviewed, "dismissed": dismissed, "case": case}


def finding_queue_status(
    finding_id: str,
    state: Optional[Dict[str, Any]] = None,
) -> str:
    """Primary queue for a finding: dismissed > case > done > open."""
    st = normalize_triage_state(state)
    fid = str(finding_id or "").strip()
    if not fid:
        return QUEUE_OPEN
    if fid in (st.get("dismissed") or {}):
        return QUEUE_DISMISSED
    if fid in (st.get("case") or []):
        return QUEUE_CASE
    if fid in (st.get("reviewed") or []):
        return QUEUE_DONE
    return QUEUE_OPEN


def filter_by_queue(
    items: Sequence[Dict[str, Any]],
    state: Optional[Dict[str, Any]] = None,
    *,
    queue: str = QUEUE_OPEN,
) -> List[Dict[str, Any]]:
    q = str(queue or QUEUE_OPEN).strip().lower() or QUEUE_OPEN
    if q not in QUEUE_IDS:
        q = QUEUE_OPEN
    st = normalize_triage_state(state)
    out: List[Dict[str, Any]] = []
    for f in items:
        if not isinstance(f, dict):
            continue
        if finding_queue_status(str(f.get("id") or ""), st) == q:
            out.append(f)
    return out


def queue_counts(
    items: Sequence[Dict[str, Any]],
    state: Optional[Dict[str, Any]] = None,
) -> Dict[str, int]:
    st = normalize_triage_state(state)
    counts = {qid: 0 for qid in QUEUE_IDS}
    for f in items:
        if not isinstance(f, dict):
            continue
        fid = str(f.get("id") or "").strip()
        if not fid:
            # Findings without ids stay in Open so they remain reachable.
            counts[QUEUE_OPEN] += 1
            continue
        counts[finding_queue_status(fid, st)] += 1
    return counts


def apply_triage_action(
    state: Optional[Dict[str, Any]],
    finding_id: str,
    action: str,
    *,
    reason: str = "",
) -> Dict[str, Any]:
    out = normalize_triage_state(state)
    fid = str(finding_id or "").strip()
    act = str(action or "").strip().lower()
    if not fid:
        return out
    if act in ("reviewed", "done"):
        reviewed = list(out.get("reviewed") or [])
        if fid not in reviewed:
            reviewed.append(fid)
        out["reviewed"] = reviewed
        # Leaving dismissed when marking Done keeps the queue unambiguous.
        dismissed = dict(out.get("dismissed") or {})
        dismissed.pop(fid, None)
        out["dismissed"] = dismissed
    elif act in ("unreviewed", "undo_done"):
        out["reviewed"] = [x for x in (out.get("reviewed") or []) if x != fid]
    elif act == "dismiss":
        dismissed = dict(out.get("dismissed") or {})
        dismissed[fid] = str(reason or "Dismissed")
        out["dismissed"] = dismissed
        out["reviewed"] = [x for x in (out.get("reviewed") or []) if x != fid]
        out["case"] = [x for x in (out.get("case") or []) if x != fid]
    elif act in ("undismiss", "restore"):
        dismissed = dict(out.get("dismissed") or {})
        dismissed.pop(fid, None)
        out["dismissed"] = dismissed
    elif act == "case":
        case = list(out.get("case") or [])
        if fid not in case:
            case.append(fid)
        out["case"] = case
        dismissed = dict(out.get("dismissed") or {})
        dismissed.pop(fid, None)
        out["dismissed"] = dismissed
    elif act == "uncase":
        out["case"] = [x for x in (out.get("case") or []) if x != fid]
    return out


def format_triage_audit_text(
    findings: Sequence[Dict[str, Any]],
    state: Optional[Dict[str, Any]] = None,
) -> str:
    """Append Done / Case / Dismissed sections for Save as text."""
    st = normalize_triage_state(state)
    by_id = {
        str(f.get("id") or "").strip(): f
        for f in findings
        if isinstance(f, dict) and str(f.get("id") or "").strip()
    }
    lines: List[str] = []

    def _title(fid: str) -> str:
        f = by_id.get(fid) or {}
        return str(f.get("title") or fid)

    done = list(st.get("reviewed") or [])
    case = list(st.get("case") or [])
    dismissed = dict(st.get("dismissed") or {})
    if done:
        lines.append("Done:")
        for fid in done:
            lines.append(f"  - {_title(fid)} (id={fid})")
        lines.append("")
    if case:
        lines.append("In case:")
        for fid in case:
            lines.append(f"  - {_title(fid)} (id={fid})")
        lines.append("")
    if dismissed:
        lines.append("Dismissed:")
        for fid, reason in dismissed.items():
            lines.append(f"  - {_title(fid)} (id={fid}): {reason}")
        lines.append("")
    return "\n".join(lines).rstrip()
